# Collecting Static Datasets
1. USA county population
2. USA state abbreviations
3. Hospital capacity in USA
4. Geographical boundaries of US states
5. Geographical boundaries of US counties
6. Geographical locations of US counties

In [36]:
!wget -P Data/  https://covid19-lake.s3.us-east-2.amazonaws.com/static-datasets/csv/CountyPopulation/County_Population.csv

--2020-04-17 11:59:01--  https://covid19-lake.s3.us-east-2.amazonaws.com/static-datasets/csv/CountyPopulation/County_Population.csv
Resolving covid19-lake.s3.us-east-2.amazonaws.com... 52.219.100.32
Connecting to covid19-lake.s3.us-east-2.amazonaws.com|52.219.100.32|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 146906 (143K) [text/csv]
Saving to: 'Data/County_Population.csv'

County_Population.c 100%[===================>] 143.46K   519KB/s    in 0.3s    

2020-04-17 11:59:02 (519 KB/s) - 'Data/County_Population.csv' saved [146906/146906]



In [37]:
!wget -P Data/  https://covid19-lake.s3.us-east-2.amazonaws.com/static-datasets/csv/state-abv/states_abv.csv

--2020-04-17 11:59:04--  https://covid19-lake.s3.us-east-2.amazonaws.com/static-datasets/csv/state-abv/states_abv.csv
Resolving covid19-lake.s3.us-east-2.amazonaws.com... 52.219.100.32
Connecting to covid19-lake.s3.us-east-2.amazonaws.com|52.219.100.32|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 665 [text/csv]
Saving to: 'Data/states_abv.csv'

states_abv.csv      100%[===================>]     665  --.-KB/s    in 0s      

2020-04-17 11:59:04 (18.7 MB/s) - 'Data/states_abv.csv' saved [665/665]



In [40]:
!wget -P Data/ https://covid19-lake.s3.us-east-2.amazonaws.com/rearc-usa-hospital-beds/json/usa-hospital-beds.geojson

--2020-04-17 12:05:46--  https://covid19-lake.s3.us-east-2.amazonaws.com/rearc-usa-hospital-beds/json/usa-hospital-beds.geojson
Resolving covid19-lake.s3.us-east-2.amazonaws.com... 52.219.100.104
Connecting to covid19-lake.s3.us-east-2.amazonaws.com|52.219.100.104|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4163728 (4.0M) [binary/octet-stream]
Saving to: 'Data/usa-hospital-beds.geojson'

usa-hospital-beds.g 100%[===================>]   3.97M  2.56MB/s    in 1.6s    

2020-04-17 12:05:48 (2.56 MB/s) - 'Data/usa-hospital-beds.geojson' saved [4163728/4163728]



In [55]:
!wget -P Data/ https://eric.clst.org/assets/wiki/uploads/Stuff/gz_2010_us_040_00_500k.json

--2020-04-17 13:06:37--  https://eric.clst.org/assets/wiki/uploads/Stuff/gz_2010_us_040_00_500k.json
Resolving eric.clst.org... 69.163.226.135
Connecting to eric.clst.org|69.163.226.135|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2490848 (2.4M) [application/json]
Saving to: 'Data/gz_2010_us_040_00_500k.json'

gz_2010_us_040_00_5 100%[===================>]   2.38M  1.99MB/s    in 1.2s    

2020-04-17 13:06:39 (1.99 MB/s) - 'Data/gz_2010_us_040_00_500k.json' saved [2490848/2490848]



In [56]:
!wget -P Data/ https://eric.clst.org/assets/wiki/uploads/Stuff/gz_2010_us_050_00_500k.json

--2020-04-17 13:06:40--  https://eric.clst.org/assets/wiki/uploads/Stuff/gz_2010_us_050_00_500k.json
Resolving eric.clst.org... 69.163.226.135
Connecting to eric.clst.org|69.163.226.135|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23543838 (22M) [application/json]
Saving to: 'Data/gz_2010_us_050_00_500k.json'

gz_2010_us_050_00_5 100%[===================>]  22.45M  2.56MB/s    in 9.2s    

2020-04-17 13:06:50 (2.45 MB/s) - 'Data/gz_2010_us_050_00_500k.json' saved [23543838/23543838]



In [ ]:
!wget -P Data/ https://www2.census.gov/geo/docs/maps-data/data/gazetteer/2019_Gazetteer/2019_Gaz_counties_national.zip

# Collecting Dynamic Datasets
1. Number of Covid19 cases in US counties
2. Number of Covid19 cases in US states

In [1]:
!wget -P Data/ https://covid19-lake.s3.us-east-2.amazonaws.com/rearc-covid-19-nyt-data-in-usa/csv/us-counties/us-counties.csv

--2020-04-24 11:30:19--  https://covid19-lake.s3.us-east-2.amazonaws.com/rearc-covid-19-nyt-data-in-usa/csv/us-counties/us-counties.csv
Resolving covid19-lake.s3.us-east-2.amazonaws.com... 52.219.100.24
Connecting to covid19-lake.s3.us-east-2.amazonaws.com|52.219.100.24|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3331795 (3.2M) [binary/octet-stream]
Saving to: 'Data/us-counties.csv'

us-counties.csv     100%[===================>]   3.18M  1.97MB/s    in 1.6s    

2020-04-24 11:30:21 (1.97 MB/s) - 'Data/us-counties.csv' saved [3331795/3331795]



In [2]:
!wget -P Data/ https://covid19-lake.s3.us-east-2.amazonaws.com/rearc-covid-19-nyt-data-in-usa/csv/us-states/us-states.csv

--2020-04-24 11:30:24--  https://covid19-lake.s3.us-east-2.amazonaws.com/rearc-covid-19-nyt-data-in-usa/csv/us-states/us-states.csv
Resolving covid19-lake.s3.us-east-2.amazonaws.com... 52.219.100.24
Connecting to covid19-lake.s3.us-east-2.amazonaws.com|52.219.100.24|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 85920 (84K) [binary/octet-stream]
Saving to: 'Data/us-states.csv'

us-states.csv       100%[===================>]  83.91K  --.-KB/s    in 0.1s    

2020-04-24 11:30:25 (586 KB/s) - 'Data/us-states.csv' saved [85920/85920]



# Importing libraries

In [111]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

import json
from pandas.io.json import json_normalize 

from sklearn.cluster import KMeans
from sklearn.cluster import DBSCAN 
from sklearn import preprocessing
import sklearn.utils
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

%matplotlib inline 
import matplotlib.cm as cm
import matplotlib.colors as colors
import matplotlib as mpl
import matplotlib.pyplot as plt

import seaborn as sns

#!conda install -c conda-forge folium=0.5.0 --yes
import folium

#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim

# Data cleaning

# County population data

In [112]:
US_pop = pd.read_csv('Data/County_Population.csv')
US_pop.head()

Id   Id2   County    State  Population Estimate 2018
0  0500000US01001  1001  Autauga  Alabama                     55601
1  0500000US01003  1003  Baldwin  Alabama                    218022
2  0500000US01005  1005  Barbour  Alabama                     24881
3  0500000US01007  1007     Bibb  Alabama                     22400
4  0500000US01009  1009   Blount  Alabama                     57840

__Fixing nonalpha characters__

In [113]:
US_pop[US_pop['County'].str.contains('Ana')]

Id    Id2    County       State  Population Estimate 2018
1802  0500000US35013  35013  Do�a Ana  New Mexico                    217522

In [114]:
US_pop.at[1802, 'County'] = 'Dona Ana'

In [115]:
US_pop.count()

Id                          3220
Id2                         3220
County                      3220
State                       3220
Population Estimate 2018    3220
dtype: int64

__State abbreviation data of 51 states (main land, doesn't include Puerto Rico, etc...)__

In [116]:
state_abb = pd.read_csv('Data/states_abv.csv')
state_abb.head()

State Abbreviation
0     Alabama           AL
1      Alaska           AK
2     Arizona           AZ
3    Arkansas           AR
4  California           CA

In [117]:
state_abb.count()

State           51
Abbreviation    51
dtype: int64

__checking for missing data when joining population and state abbreviation data__

In [118]:
US_pop.join(state_abb.set_index('State'), on = 'State').count()

Id                          3220
Id2                         3220
County                      3220
State                       3220
Population Estimate 2018    3220
Abbreviation                3142
dtype: int64

__finding those states without abbreviation data__

In [119]:
US_pop[US_pop.join(state_abb.set_index('State'), on = 'State')['Abbreviation'].isna()]['State'].unique()

array(['Puerto Rico'], dtype=object)

__Joining population and state abbreviation data__

In [120]:
US_pop_c = US_pop.join(state_abb.set_index('State'), on = 'State').dropna().reset_index(drop=True)
US_pop_c

Id    Id2      County    State  Population Estimate 2018  \
0     0500000US01001   1001     Autauga  Alabama                     55601   
1     0500000US01003   1003     Baldwin  Alabama                    218022   
2     0500000US01005   1005     Barbour  Alabama                     24881   
3     0500000US01007   1007        Bibb  Alabama                     22400   
4     0500000US01009   1009      Blount  Alabama                     57840   
...              ...    ...         ...      ...                       ...   
3137  0500000US56037  56037  Sweetwater  Wyoming                     43051   
3138  0500000US56039  56039       Teton  Wyoming                     23081   
3139  0500000US56041  56041       Uinta  Wyoming                     20299   
3140  0500000US56043  56043    Washakie  Wyoming                      7885   
3141  0500000US56045  56045      Weston  Wyoming                      6967   

     Abbreviation  
0              AL  
1              AL  
2              AL  
3              AL  
4              AL  
...           ...  
3137           WY  
3138           WY  
3139           WY  
3140           WY  
3141           WY  

[3142 rows x 6 columns]

# Hospital capacity in each county data

In [121]:
US_hosp = pd.read_json('Data/usa-hospital-beds.geojson', lines=True)
US_hosp.head()

OBJECTID                                      HOSPITAL_NAME  \
0      3001  Western Reserve Hospital (FKA Summa Western Re...   
1      3002                 Summa Health System - Akron Campus   
2      3003             Summa Health System - Barberton Campus   
3      3004                                      Akron General   
4      3005                  Crystal Clinic Orthopaedic Center   

                    HOSPITAL_TYPE           HQ_ADDRESS HQ_ADDRESS1  \
0  Short Term Acute Care Hospital         1900 23rd St        None   
1  Short Term Acute Care Hospital      525 E Market St        None   
2  Short Term Acute Care Hospital        155 5th St Ne        None   
3  Short Term Acute Care Hospital  1 Akron General Ave        None   
4  Short Term Acute Care Hospital        444 N Main St        None   

          HQ_CITY HQ_STATE  HQ_ZIP_CODE COUNTY_NAME STATE_NAME  ...  \
0  Cuyahoga Falls       OH      44223.0      Summit       Ohio  ...   
1           Akron       OH      44304.0      Summit       Ohio  ...   
2       Barberton       OH      44203.0      Summit       Ohio  ...   
3           Akron       OH      44307.0      Summit       Ohio  ...   
4           Akron       OH      44310.0      Summit       Ohio  ...   

   NUM_LICENSED_BEDS  NUM_STAFFED_BEDS  NUM_ICU_BEDS  ADULT_ICU_BEDS  \
0              272.0              83.0            12              12   
1              649.0             584.0            64              64   
2              271.0             158.0            12              12   
3              383.0             383.0            34              34   
4               94.0              59.0            10              10   

   PEDI_ICU_BEDS  BED_UTILIZATION  Potential_Increase_In_Bed_Capac  \
0            0.0         0.430764                              189   
1            0.0         0.540098                               65   
2            0.0         0.628819                              113   
3            0.0         0.706661                                0   
4            0.0         0.431066                               35   

   AVG_VENTILATOR_USAGE   latitude  longtitude  
0                   2.0  41.133034  -81.510687  
1                  30.0  41.080189  -81.500489  
2                   0.0  41.014805  -81.589456  
3                  10.0  41.078325  -81.532470  
4                   2.0  41.096977  -81.513552  

[5 rows x 23 columns]

In [122]:
US_hosp.count()

OBJECTID                           6618
HOSPITAL_NAME                      6618
HOSPITAL_TYPE                      6618
HQ_ADDRESS                         6618
HQ_ADDRESS1                         364
HQ_CITY                            6618
HQ_STATE                           6618
HQ_ZIP_CODE                        6617
COUNTY_NAME                        6593
STATE_NAME                         6593
STATE_FIPS                         6593
CNTY_FIPS                          6593
FIPS                               6593
NUM_LICENSED_BEDS                  6402
NUM_STAFFED_BEDS                   6327
NUM_ICU_BEDS                       6618
ADULT_ICU_BEDS                     6618
PEDI_ICU_BEDS                      5902
BED_UTILIZATION                    5905
Potential_Increase_In_Bed_Capac    6618
AVG_VENTILATOR_USAGE               6572
latitude                           6600
longtitude                         6600
dtype: int64

__removing PR as we removed it in the US_pop_c__

In [123]:
US_hosp_c = US_hosp.drop(['OBJECTID', 'HOSPITAL_TYPE' , 'HQ_ADDRESS', 'HQ_ADDRESS1', 'HQ_CITY', 'HQ_ZIP_CODE', 'STATE_NAME' , 'STATE_FIPS' , 'CNTY_FIPS', 'FIPS', 'latitude', 'longtitude'], axis = 1).reset_index(drop=True)
US_hosp_c = US_hosp_c[US_hosp_c['HQ_STATE'] != 'PR']
US_hosp_c.reset_index(drop=True)

HOSPITAL_NAME HQ_STATE COUNTY_NAME  \
0     Western Reserve Hospital (FKA Summa Western Re...       OH      Summit   
1                    Summa Health System - Akron Campus       OH      Summit   
2                Summa Health System - Barberton Campus       OH      Summit   
3                                         Akron General       OH      Summit   
4                     Crystal Clinic Orthopaedic Center       OH      Summit   
...                                                 ...      ...         ...   
6546                           United Hospital District       MN   Faribault   
6547                  Mayo Clinic Hospital - Albert Lea       MN    Freeborn   
6548  Mayo Clinic Hospital - Red Wing (FKA Fairview ...       MN     Goodhue   
6549                Mayo Clinic Hospital - Cannon Falls       MN     Goodhue   
6550         Prairie Ridge Hospital and Health Services       MN       Grant   

      NUM_LICENSED_BEDS  NUM_STAFFED_BEDS  NUM_ICU_BEDS  ADULT_ICU_BEDS  \
0                 272.0              83.0            12              12   
1                 649.0             584.0            64              64   
2                 271.0             158.0            12              12   
3                 383.0             383.0            34              34   
4                  94.0              59.0            10              10   
...                 ...               ...           ...             ...   
6546               43.0              24.0             4               4   
6547              159.0             159.0            18              18   
6548               50.0              50.0             4               4   
6549               15.0              15.0             5               5   
6550               10.0              10.0             2               2   

      PEDI_ICU_BEDS  BED_UTILIZATION  Potential_Increase_In_Bed_Capac  \
0               0.0         0.430764                              189   
1               0.0         0.540098                               65   
2               0.0         0.628819                              113   
3               0.0         0.706661                                0   
4               0.0         0.431066                               35   
...             ...              ...                              ...   
6546            0.0         0.355594                               19   
6547            0.0         0.251624                                0   
6548            0.0         0.256767                                0   
6549            0.0         0.608950                                0   
6550            0.0         0.171781                                0   

      AVG_VENTILATOR_USAGE  
0                      2.0  
1                     30.0  
2                      0.0  
3                     10.0  
4                      2.0  
...                    ...  
6546                   1.0  
6547                   4.0  
6548                   2.0  
6549                   2.0  
6550                   0.0  

[6551 rows x 11 columns]

__joining the hospital data with US_pop_c using county names__

In [124]:
US_hosp_pop = US_hosp_c.join(US_pop_c.set_index(['County', 'Abbreviation']), on = ['COUNTY_NAME', 'HQ_STATE']).drop('State', axis = 1).reset_index(drop = True)
US_hosp_pop.count()

HOSPITAL_NAME                      6551
HQ_STATE                           6551
COUNTY_NAME                        6543
NUM_LICENSED_BEDS                  6344
NUM_STAFFED_BEDS                   6269
NUM_ICU_BEDS                       6551
ADULT_ICU_BEDS                     6551
PEDI_ICU_BEDS                      5847
BED_UTILIZATION                    5849
Potential_Increase_In_Bed_Capac    6551
AVG_VENTILATOR_USAGE               6505
Id                                 6172
Id2                                6172
Population Estimate 2018           6172
dtype: int64

__There are 110 counties in the hospital data that we couldn't find a match for it in the US_pop_c__

In [125]:
US_hosp_pop[US_hosp_pop['Population Estimate 2018'].isna()].groupby(by = ['COUNTY_NAME', 'HQ_STATE']).count().index.size

110

__Let's get these 110 counties__

In [126]:
temp = US_hosp_pop[US_hosp_pop['Population Estimate 2018'].isna()].groupby(by = ['COUNTY_NAME', 'HQ_STATE']).count().index.tolist()
temp

[('Acadia', 'LA'),
 ('Alexandria', 'VA'),
 ('Allen', 'LA'),
 ('Anchorage', 'AK'),
 ('Ascension', 'LA'),
 ('Assumption', 'LA'),
 ('Avoyelles', 'LA'),
 ('Baltimore City', 'MD'),
 ('Baltimore County', 'MD'),
 ('Beauregard', 'LA'),
 ('Bedford County', 'VA'),
 ('Bethel', 'AK'),
 ('Bienville', 'LA'),
 ('Bossier', 'LA'),
 ('Bristol', 'VA'),
 ('Caddo', 'LA'),
 ('Calcasieu', 'LA'),
 ('Caldwell', 'LA'),
 ('Cameron', 'LA'),
 ('Charlottesville', 'VA'),
 ('Chesapeake', 'VA'),
 ('Claiborne', 'LA'),
 ('Concordia', 'LA'),
 ('Danville', 'VA'),
 ('De Soto', 'LA'),
 ('Dillingham', 'AK'),
 ('East Baton Rouge', 'LA'),
 ('East Carroll', 'LA'),
 ('East Feliciana', 'LA'),
 ('Emporia', 'VA'),
 ('Evangeline', 'LA'),
 ('Fairbanks North Star', 'AK'),
 ('Fairfax County', 'VA'),
 ('Franklin', 'LA'),
 ('Franklin City', 'VA'),
 ('Franklin County', 'VA'),
 ('Fredericksburg', 'VA'),
 ('Galax', 'VA'),
 ('Hampton', 'VA'),
 ('Hopewell', 'VA'),
 ('Iberia', 'LA'),
 ('Iberville', 'LA'),
 ('Jackson', 'LA'),
 ('Jefferson', 'LA

__This is a brute force way to find similar names of the 110 counties in the US_pop_c data and then replace them in the hosptial data__

In [127]:
count_not_found = 0
new_counties = []
old_counties = []
for i in temp:
    old_counties.append(i[0])
    
    new_county = US_pop_c[US_pop_c['County'].str.contains(i[0]) & US_pop_c['Abbreviation'].str.contains(i[1])]['County'].values
    if not new_county.any(): 
        new_county = US_pop_c[US_pop_c['County'].str.contains(i[0].replace(' County', '')) & US_pop_c['Abbreviation'].str.contains(i[1])]['County'].values
        if not new_county.any():    
            new_county = US_pop_c[US_pop_c['County'].str.contains(i[0].replace(' County', '').capitalize()) & US_pop_c['Abbreviation'].str.contains(i[1])]['County'].values
            if not new_county.any():
                new_county = US_pop_c[US_pop_c['County'].str.contains(i[0].replace(' City', ' city')) & US_pop_c['Abbreviation'].str.contains(i[1])]['County'].values
                if not new_county.any():
                    new_county = US_pop_c[US_pop_c['County'].str.contains(i[0].replace(' ', '')) & US_pop_c['Abbreviation'].str.contains(i[1])]['County'].values
    
    
    if new_county.any(): 
        new_county_name = new_county[0]
        new_counties.append(new_county_name)
        US_hosp_pop[(US_hosp_pop['COUNTY_NAME'] == i[0]) & (US_hosp_pop['HQ_STATE'] == i[1])] = US_hosp_pop[(US_hosp_pop['COUNTY_NAME'] == i[0]) & (US_hosp_pop['HQ_STATE'] == i[1])].replace({'COUNTY_NAME' : {i[0] : new_county_name}})
    else:
        new_county_name = '>>>>>>>>>>> NOT FOUND: ' + i[0] + ' ' + i[1]
        count_not_found += 1
    print(i[0] + ' >>>>> ' + new_county_name)
print(count_not_found)
print(len(new_counties))
print(len(old_counties))

Acadia >>>>> Acadia Parish
Alexandria >>>>> Alexandria city
Allen >>>>> Allen Parish
Anchorage >>>>> Anchorage Municipality
Ascension >>>>> Ascension Parish
Assumption >>>>> Assumption Parish
Avoyelles >>>>> Avoyelles Parish
Baltimore City >>>>> Baltimore city
Baltimore County >>>>> Baltimore
Beauregard >>>>> Beauregard Parish
Bedford County >>>>> Bedford
Bethel >>>>> Bethel Census Area
Bienville >>>>> Bienville Parish
Bossier >>>>> Bossier Parish
Bristol >>>>> Bristol city
Caddo >>>>> Caddo Parish
Calcasieu >>>>> Calcasieu Parish
Caldwell >>>>> Caldwell Parish
Cameron >>>>> Cameron Parish
Charlottesville >>>>> Charlottesville city
Chesapeake >>>>> Chesapeake city
Claiborne >>>>> Claiborne Parish
Concordia >>>>> Concordia Parish
Danville >>>>> Danville city
De Soto >>>>> De Soto Parish
Dillingham >>>>> Dillingham Census Area
East Baton Rouge >>>>> East Baton Rouge Parish
East Carroll >>>>> East Carroll Parish
East Feliciana >>>>> East Feliciana Parish
Emporia >>>>> Emporia city
Evangel

__now we join them again, we see that we get population data for all the counties. Note: there are hosptials without known county, so we will drop thoses__

In [128]:
US_hosp_pop.drop(['Population Estimate 2018', 'Id', 'Id2'], axis = 1, inplace = True)
US_hosp_pop_c = US_hosp_pop.join(US_pop_c.set_index(['County', 'Abbreviation']), on = ['COUNTY_NAME', 'HQ_STATE']).drop('State', axis = 1).reset_index(drop = True)
US_hosp_pop_c.count()

HOSPITAL_NAME                      6551
HQ_STATE                           6551
COUNTY_NAME                        6543
NUM_LICENSED_BEDS                  6344
NUM_STAFFED_BEDS                   6269
NUM_ICU_BEDS                       6551
ADULT_ICU_BEDS                     6551
PEDI_ICU_BEDS                      5847
BED_UTILIZATION                    5849
Potential_Increase_In_Bed_Capac    6551
AVG_VENTILATOR_USAGE               6505
Id                                 6543
Id2                                6543
Population Estimate 2018           6543
dtype: int64

__Final check to see if we have any missing values__

In [129]:
US_hosp_pop_c = US_hosp_pop_c[US_hosp_pop_c['Population Estimate 2018'].notna()]
US_hosp_pop_c.fillna(value = 0, inplace = True)
US_hosp_pop_c.count()

HOSPITAL_NAME                      6543
HQ_STATE                           6543
COUNTY_NAME                        6543
NUM_LICENSED_BEDS                  6543
NUM_STAFFED_BEDS                   6543
NUM_ICU_BEDS                       6543
ADULT_ICU_BEDS                     6543
PEDI_ICU_BEDS                      6543
BED_UTILIZATION                    6543
Potential_Increase_In_Bed_Capac    6543
AVG_VENTILATOR_USAGE               6543
Id                                 6543
Id2                                6543
Population Estimate 2018           6543
dtype: int64

__Here we want to get the total number of beds, icu, etc... in each county, so we Sum() over the counties__

In [130]:
temp1 = US_hosp_pop_c.groupby(by=['Id', 'Id2', 'Population Estimate 2018']).sum().reset_index()
temp1

Id      Id2  Population Estimate 2018  NUM_LICENSED_BEDS  \
0     0500000US01001   1001.0                   55601.0               85.0   
1     0500000US01003   1003.0                  218022.0              386.0   
2     0500000US01005   1005.0                   24881.0               74.0   
3     0500000US01007   1007.0                   22400.0               35.0   
4     0500000US01009   1009.0                   57840.0               25.0   
...              ...      ...                       ...                ...   
2476  0500000US56037  56037.0                   43051.0              115.0   
2477  0500000US56039  56039.0                   23081.0               48.0   
2478  0500000US56041  56041.0                   20299.0              225.0   
2479  0500000US56043  56043.0                    7885.0               18.0   
2480  0500000US56045  56045.0                    6967.0               12.0   

      NUM_STAFFED_BEDS  NUM_ICU_BEDS  ADULT_ICU_BEDS  PEDI_ICU_BEDS  \
0                 55.0             6               6            0.0   
1                362.0            51              51            0.0   
2                 30.0             5               5            0.0   
3                 25.0             4               4            0.0   
4                 25.0             6               6            0.0   
...                ...           ...             ...            ...   
2476              74.0            13              13            0.0   
2477              48.0             6               6            0.0   
2478             105.0             6               6            0.0   
2479              18.0             3               3            0.0   
2480              12.0             2               2            0.0   

      BED_UTILIZATION  Potential_Increase_In_Bed_Capac  AVG_VENTILATOR_USAGE  
0            0.623113                               30                   2.0  
1            2.254794                               24                   8.0  
2            0.430640                               44                   2.0  
3            0.376219                               10                   1.0  
4            0.499068                                0                   2.0  
...               ...                              ...                   ...  
2476         0.216816                               41                   2.0  
2477         0.319578                                0                   2.0  
2478         0.901606                              120                   2.0  
2479         0.287976                                0                   2.0  
2480         0.599641                                0                   0.0  

[2481 rows x 11 columns]

__Here we want to get the total number of hospitals in each county__

In [131]:
temp2 = US_hosp_pop_c.groupby(by='Id').count()[['HOSPITAL_NAME']].reset_index()
temp2

Id  HOSPITAL_NAME
0     0500000US01001              1
1     0500000US01003              4
2     0500000US01005              1
3     0500000US01007              1
4     0500000US01009              1
...              ...            ...
2476  0500000US56037              2
2477  0500000US56039              1
2478  0500000US56041              2
2479  0500000US56043              1
2480  0500000US56045              1

[2481 rows x 2 columns]

__We join the two previos temp dataframes to get hospital capacity data for each county because we don't care about the names of the hopsitals, etc.. NOTE: Here we only have data about hospital capacity in only 2481 counties instead of the 3142 counties in US_pop_c__

In [132]:
US_hosp_pop_c = temp1.join(temp2.set_index('Id'), on = 'Id').rename(columns = {'HOSPITAL_NAME' : 'NUM_HOSPITALS'})
US_hosp_pop_c

Id      Id2  Population Estimate 2018  NUM_LICENSED_BEDS  \
0     0500000US01001   1001.0                   55601.0               85.0   
1     0500000US01003   1003.0                  218022.0              386.0   
2     0500000US01005   1005.0                   24881.0               74.0   
3     0500000US01007   1007.0                   22400.0               35.0   
4     0500000US01009   1009.0                   57840.0               25.0   
...              ...      ...                       ...                ...   
2476  0500000US56037  56037.0                   43051.0              115.0   
2477  0500000US56039  56039.0                   23081.0               48.0   
2478  0500000US56041  56041.0                   20299.0              225.0   
2479  0500000US56043  56043.0                    7885.0               18.0   
2480  0500000US56045  56045.0                    6967.0               12.0   

      NUM_STAFFED_BEDS  NUM_ICU_BEDS  ADULT_ICU_BEDS  PEDI_ICU_BEDS  \
0                 55.0             6               6            0.0   
1                362.0            51              51            0.0   
2                 30.0             5               5            0.0   
3                 25.0             4               4            0.0   
4                 25.0             6               6            0.0   
...                ...           ...             ...            ...   
2476              74.0            13              13            0.0   
2477              48.0             6               6            0.0   
2478             105.0             6               6            0.0   
2479              18.0             3               3            0.0   
2480              12.0             2               2            0.0   

      BED_UTILIZATION  Potential_Increase_In_Bed_Capac  AVG_VENTILATOR_USAGE  \
0            0.623113                               30                   2.0   
1            2.254794                               24                   8.0   
2            0.430640                               44                   2.0   
3            0.376219                               10                   1.0   
4            0.499068                                0                   2.0   
...               ...                              ...                   ...   
2476         0.216816                               41                   2.0   
2477         0.319578                                0                   2.0   
2478         0.901606                              120                   2.0   
2479         0.287976                                0                   2.0   
2480         0.599641                                0                   0.0   

      NUM_HOSPITALS  
0                 1  
1                 4  
2                 1  
3                 1  
4                 1  
...             ...  
2476              2  
2477              1  
2478              2  
2479              1  
2480              1  

[2481 rows x 12 columns]

__Correcting the bed utilization which is a percentage__

In [133]:
US_hosp_pop_c['BED_UTILIZATION'] = US_hosp_pop_c['BED_UTILIZATION']/US_hosp_pop_c['NUM_HOSPITALS']
US_hosp_pop_c

Id      Id2  Population Estimate 2018  NUM_LICENSED_BEDS  \
0     0500000US01001   1001.0                   55601.0               85.0   
1     0500000US01003   1003.0                  218022.0              386.0   
2     0500000US01005   1005.0                   24881.0               74.0   
3     0500000US01007   1007.0                   22400.0               35.0   
4     0500000US01009   1009.0                   57840.0               25.0   
...              ...      ...                       ...                ...   
2476  0500000US56037  56037.0                   43051.0              115.0   
2477  0500000US56039  56039.0                   23081.0               48.0   
2478  0500000US56041  56041.0                   20299.0              225.0   
2479  0500000US56043  56043.0                    7885.0               18.0   
2480  0500000US56045  56045.0                    6967.0               12.0   

      NUM_STAFFED_BEDS  NUM_ICU_BEDS  ADULT_ICU_BEDS  PEDI_ICU_BEDS  \
0                 55.0             6               6            0.0   
1                362.0            51              51            0.0   
2                 30.0             5               5            0.0   
3                 25.0             4               4            0.0   
4                 25.0             6               6            0.0   
...                ...           ...             ...            ...   
2476              74.0            13              13            0.0   
2477              48.0             6               6            0.0   
2478             105.0             6               6            0.0   
2479              18.0             3               3            0.0   
2480              12.0             2               2            0.0   

      BED_UTILIZATION  Potential_Increase_In_Bed_Capac  AVG_VENTILATOR_USAGE  \
0            0.623113                               30                   2.0   
1            0.563699                               24                   8.0   
2            0.430640                               44                   2.0   
3            0.376219                               10                   1.0   
4            0.499068                                0                   2.0   
...               ...                              ...                   ...   
2476         0.108408                               41                   2.0   
2477         0.319578                                0                   2.0   
2478         0.450803                              120                   2.0   
2479         0.287976                                0                   2.0   
2480         0.599641                                0                   0.0   

      NUM_HOSPITALS  
0                 1  
1                 4  
2                 1  
3                 1  
4                 1  
...             ...  
2476              2  
2477              1  
2478              2  
2479              1  
2480              1  

[2481 rows x 12 columns]

# Cases in US counties data

In [134]:
US_cases = pd.read_csv('Data/us-counties.csv')
US_cases

date      county       state     fips  cases  deaths
0      2020-01-21   Snohomish  Washington  53061.0      1       0
1      2020-01-22   Snohomish  Washington  53061.0      1       0
2      2020-01-23   Snohomish  Washington  53061.0      1       0
3      2020-01-24        Cook    Illinois  17031.0      1       0
4      2020-01-24   Snohomish  Washington  53061.0      1       0
...           ...         ...         ...      ...    ...     ...
81335  2020-04-22    Sublette     Wyoming  56035.0      1       0
81336  2020-04-22  Sweetwater     Wyoming  56037.0     10       0
81337  2020-04-22       Teton     Wyoming  56039.0     63       0
81338  2020-04-22       Uinta     Wyoming  56041.0      6       0
81339  2020-04-22    Washakie     Wyoming  56043.0      4       0

[81340 rows x 6 columns]

In [135]:
US_cases.count()

date      81340
county    81340
state     81340
fips      80324
cases     81340
deaths    81340
dtype: int64

__Getting the cases in states which is not in our interest (the 51 states in state_abb data)__

In [136]:
US_cases[US_cases.join(state_abb.set_index('State'), on = 'state')['Abbreviation'].isna()]

date   county                     state  fips  cases  deaths
1858   2020-03-13  Unknown               Puerto Rico   NaN      3       0
2220   2020-03-14  Unknown               Puerto Rico   NaN      4       0
2267   2020-03-14  Unknown            Virgin Islands   NaN      1       0
2422   2020-03-15  Unknown                      Guam   NaN      3       0
2642   2020-03-15  Unknown               Puerto Rico   NaN      5       0
...           ...      ...                       ...   ...    ...     ...
78246  2020-04-21  Unknown            Virgin Islands   NaN     51       2
79063  2020-04-22  Unknown                      Guam   NaN    911       6
80358  2020-04-22  Unknown  Northern Mariana Islands   NaN     14       2
80613  2020-04-22  Unknown               Puerto Rico   NaN    915      46
81039  2020-04-22  Unknown            Virgin Islands   NaN     51       2

[146 rows x 6 columns]

__Drop the previos states, and join the dataset with the state abbreviation data__

In [137]:
US_cases_c = US_cases.drop(['fips'], axis = 1).reset_index(drop=True)  
US_cases_c = US_cases_c.join(state_abb.set_index('State'), on = 'state').dropna()
US_cases_c.count()

date            81194
county          81194
state           81194
cases           81194
deaths          81194
Abbreviation    81194
dtype: int64

__Unfortunately there are cases in unknown counties but known states__

In [138]:
US_cases_c[US_cases_c['county'] == 'Unknown']

date   county         state  cases  deaths Abbreviation
418    2020-03-01  Unknown  Rhode Island      2       0           RI
450    2020-03-02  Unknown  Rhode Island      2       0           RI
485    2020-03-03  Unknown  Rhode Island      2       0           RI
522    2020-03-04  Unknown  Rhode Island      2       0           RI
569    2020-03-05  Unknown  Rhode Island      2       0           RI
...           ...      ...           ...    ...     ...          ...
80144  2020-04-22  Unknown    New Jersey    591       2           NJ
80172  2020-04-22  Unknown    New Mexico      1       0           NM
80618  2020-04-22  Unknown  Rhode Island   1000     161           RI
81019  2020-04-22  Unknown          Utah      0       2           UT
81035  2020-04-22  Unknown       Vermont      3       0           VT

[783 rows x 6 columns]

__Finding those states__

In [139]:
temp = US_cases_c[US_cases_c['county'] == 'Unknown']['Abbreviation'].unique().tolist()
temp

['RI',
 'NJ',
 'ME',
 'MA',
 'LA',
 'KY',
 'NV',
 'TN',
 'AR',
 'GA',
 'MO',
 'MN',
 'CA',
 'CO',
 'FL',
 'HI',
 'IL',
 'VT',
 'ID',
 'MI',
 'OH',
 'UT',
 'MS',
 'NE',
 'CT',
 'IN',
 'AK',
 'AZ',
 'NM',
 'NY',
 'PA',
 'VA',
 'NH',
 'SD',
 'WA',
 'DE',
 'KS',
 'ND',
 'MD',
 'IA']

__To solve the problem of the missing county names, we make an assumption: we assume that the Covid cases happened in the counties of largest population in the corresponding state. This can be justified by the method in which Covid spread__

In [140]:
for i in temp:
    set_county = US_pop_c[US_pop_c['Abbreviation'] == i].sort_values(by=['Population Estimate 2018'], ascending = False)['County'].values[0]
    US_cases_c[(US_cases_c['Abbreviation'] == i) & (US_cases_c['county'] == 'Unknown')] = US_cases_c[(US_cases_c['Abbreviation'] == i) & (US_cases_c['county'] == 'Unknown')].replace({'county' : {'Unknown' : set_county}})

__check that we don't have any unknown counties__

In [141]:
US_cases_c[US_cases_c['county'] == 'Unknown']

Empty DataFrame
Columns: [date, county, state, cases, deaths, Abbreviation]
Index: []

__fixing non alpha char in counties__

In [142]:
US_cases_c[US_cases_c['county'] == 'Doña Ana'] = US_cases_c[US_cases_c['county'] == 'Doña Ana'].replace({'county' : {'Doña Ana' : 'Dona Ana'}})

__Now, as we did before, let's check the counties name that doesn't have a match in the US_pop_c data__

In [143]:
temp1 = US_cases_c[['county', 'Abbreviation']].drop_duplicates().values
temp2 = US_pop_c[['County', 'Abbreviation']].drop_duplicates().values

odd_counties = []
found = 0
for i in temp1:
    for j in temp2:
        if (i[0] == j[0]) & (i[1] == j[1]): 
            found = 1
    
    if found == 0:        
        odd_counties.append(tuple(i))
    
    found = 0

__There are 67 counties__

In [144]:
len(set(odd_counties))

67

In [145]:
temp = set(odd_counties)
temp

{('Acadia', 'LA'),
 ('Allen', 'LA'),
 ('Anchorage', 'AK'),
 ('Ascension', 'LA'),
 ('Assumption', 'LA'),
 ('Avoyelles', 'LA'),
 ('Beauregard', 'LA'),
 ('Bienville', 'LA'),
 ('Bossier', 'LA'),
 ('Caddo', 'LA'),
 ('Calcasieu', 'LA'),
 ('Caldwell', 'LA'),
 ('Cameron', 'LA'),
 ('Catahoula', 'LA'),
 ('Claiborne', 'LA'),
 ('Concordia', 'LA'),
 ('De Soto', 'LA'),
 ('East Baton Rouge', 'LA'),
 ('East Carroll', 'LA'),
 ('East Feliciana', 'LA'),
 ('Evangeline', 'LA'),
 ('Franklin', 'LA'),
 ('Grant', 'LA'),
 ('Iberia', 'LA'),
 ('Iberville', 'LA'),
 ('Jackson', 'LA'),
 ('Jefferson', 'LA'),
 ('Jefferson Davis', 'LA'),
 ('Kansas City', 'MO'),
 ('LaSalle', 'LA'),
 ('Lafayette', 'LA'),
 ('Lafourche', 'LA'),
 ('Lincoln', 'LA'),
 ('Livingston', 'LA'),
 ('Madison', 'LA'),
 ('Morehouse', 'LA'),
 ('Natchitoches', 'LA'),
 ('New York City', 'NY'),
 ('Orleans', 'LA'),
 ('Ouachita', 'LA'),
 ('Plaquemines', 'LA'),
 ('Pointe Coupee', 'LA'),
 ('Rapides', 'LA'),
 ('Red River', 'LA'),
 ('Richland', 'LA'),
 ('Sabine'

__As we did before, we use brute force to find matching name of the counties in Pop_US_c and replace it in the US_cases_c__

In [146]:
count_not_found = 0
new_counties = []
old_counties = []
for i in temp:
    old_counties.append(i[0])
    
    new_county = US_pop_c[US_pop_c['County'].str.contains(i[0]) & US_pop_c['Abbreviation'].str.contains(i[1])]['County'].values
    if (i[0] == 'New York City'):
        new_county = US_pop_c[US_pop_c['County'].str.contains('New York') & US_pop_c['Abbreviation'].str.contains(i[1])]['County'].values
    elif (i[0] == 'Kansas City'):
        new_county = US_pop_c[US_pop_c['County'].str.contains('Jackson') & US_pop_c['Abbreviation'].str.contains(i[1])]['County'].values

        
            
    
    if new_county.any(): 
        new_county_name = new_county[0]
        new_counties.append(new_county_name)
        US_cases_c[(US_cases_c['county'] == i[0]) & (US_cases_c['Abbreviation'] == i[1])] = US_cases_c[(US_cases_c['county'] == i[0]) & (US_cases_c['Abbreviation'] == i[1])].replace({'county' : {i[0] : new_county_name}})
    
    else:
        new_county_name = '>>>>>>>>>>> NOT FOUND: ' + i[0] + ' ' + i[1]
        count_not_found += 1
    print(i[0] + ' >>>>> ' + new_county_name)
print(count_not_found)
print(len(new_counties))
print(len(old_counties))

Vermilion >>>>> Vermilion Parish
Lafayette >>>>> Lafayette Parish
Evangeline >>>>> Evangeline Parish
Winn >>>>> Winn Parish
Cameron >>>>> Cameron Parish
Calcasieu >>>>> Calcasieu Parish
Vernon >>>>> Vernon Parish
Tangipahoa >>>>> Tangipahoa Parish
Richland >>>>> Richland Parish
Orleans >>>>> Orleans Parish
Red River >>>>> Red River Parish
St. John the Baptist >>>>> St. John the Baptist Parish
Washington >>>>> Washington Parish
St. Landry >>>>> St. Landry Parish
Tensas >>>>> Tensas Parish
West Baton Rouge >>>>> West Baton Rouge Parish
Morehouse >>>>> Morehouse Parish
St. Martin >>>>> St. Martin Parish
Natchitoches >>>>> Natchitoches Parish
Iberia >>>>> Iberia Parish
Catahoula >>>>> Catahoula Parish
West Feliciana >>>>> West Feliciana Parish
Rapides >>>>> Rapides Parish
New York City >>>>> New York
Beauregard >>>>> Beauregard Parish
Assumption >>>>> Assumption Parish
Caddo >>>>> Caddo Parish
St. Helena >>>>> St. Helena Parish
De Soto >>>>> De Soto Parish
St. Mary >>>>> St. Mary Parish
Ea

__Check again that we don't have any unmatching county names__

In [147]:
temp1 = US_cases_c[['county', 'Abbreviation']].drop_duplicates().values
temp2 = US_pop_c[['County', 'Abbreviation']].drop_duplicates().values

odd_counties = []
found = 0
for i in temp1:
    for j in temp2:
        if (i[0] == j[0]) & (i[1] == j[1]): 
            found = 1
    
    if found == 0:        
        odd_counties.append(tuple(i))
    
    found = 0
    
set(odd_counties)

set()

In [148]:
US_cases_c

date      county       state  cases  deaths Abbreviation
0      2020-01-21   Snohomish  Washington      1       0           WA
1      2020-01-22   Snohomish  Washington      1       0           WA
2      2020-01-23   Snohomish  Washington      1       0           WA
3      2020-01-24        Cook    Illinois      1       0           IL
4      2020-01-24   Snohomish  Washington      1       0           WA
...           ...         ...         ...    ...     ...          ...
81335  2020-04-22    Sublette     Wyoming      1       0           WY
81336  2020-04-22  Sweetwater     Wyoming     10       0           WY
81337  2020-04-22       Teton     Wyoming     63       0           WY
81338  2020-04-22       Uinta     Wyoming      6       0           WY
81339  2020-04-22    Washakie     Wyoming      4       0           WY

[81194 rows x 6 columns]

__In the US_cases_c dataframe, there are multiple reported casese in the same day for some counties. Let's find those duplicates__

In [149]:
US_cases_c[US_cases_c[['date','county', 'Abbreviation']].duplicated()]

date      county         state  cases  deaths Abbreviation
1511   2020-03-12      Bergen    New Jersey      1       0           NJ
1804   2020-03-13      Bergen    New Jersey      1       0           NJ
2151   2020-03-14      Bergen    New Jersey      1       0           NJ
2559   2020-03-15      Bergen    New Jersey      1       0           NJ
2929   2020-03-16  Cumberland         Maine      1       0           ME
...           ...         ...           ...    ...     ...          ...
80144  2020-04-22      Bergen    New Jersey    591       2           NJ
80172  2020-04-22  Bernalillo    New Mexico      1       0           NM
80618  2020-04-22  Providence  Rhode Island   1000     161           RI
81019  2020-04-22   Salt Lake          Utah      0       2           UT
81035  2020-04-22  Chittenden       Vermont      3       0           VT

[787 rows x 6 columns]

__An example of the first duplicate. We see that on the same day, there are two numbers of reported cases. So we need to sum up those so that we have a total number of cases for each county per day.__

In [150]:
US_cases_c[(US_cases_c['date'] == '2020-03-12') & (US_cases_c['county'] == 'Bergen') & (US_cases_c['Abbreviation'] == 'NJ')]

date  county       state  cases  deaths Abbreviation
1501  2020-03-12  Bergen  New Jersey     13       1           NJ
1511  2020-03-12  Bergen  New Jersey      1       0           NJ

__Let's group by their date and counties and then sum up the cases and deaths__

In [151]:
US_cases_c = US_cases_c.groupby(by=['date', 'county', 'state', 'Abbreviation']).sum().reset_index()
US_cases_c

date                     county       state Abbreviation  cases  \
0      2020-01-21                  Snohomish  Washington           WA      1   
1      2020-01-22                  Snohomish  Washington           WA      1   
2      2020-01-23                  Snohomish  Washington           WA      1   
3      2020-01-24                       Cook    Illinois           IL      1   
4      2020-01-24                  Snohomish  Washington           WA      1   
...           ...                        ...         ...          ...    ...   
80402  2020-04-22  Yukon-Koyukuk Census Area      Alaska           AK      1   
80403  2020-04-22                       Yuma     Arizona           AZ     28   
80404  2020-04-22                       Yuma    Colorado           CO      4   
80405  2020-04-22                     Zapata       Texas           TX      7   
80406  2020-04-22                     Zavala       Texas           TX      1   

       deaths  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
80402       0  
80403       1  
80404       0  
80405       0  
80406       0  

[80407 rows x 6 columns]

__Now it is done, let's check the duplicates and confirm that they were summed up__

In [152]:
US_cases_c[(US_cases_c['date'] == '2020-03-12') & (US_cases_c['county'] == 'Bergen') & (US_cases_c['Abbreviation'] == 'NJ')]

date  county       state Abbreviation  cases  deaths
1379  2020-03-12  Bergen  New Jersey           NJ     14       1

In [153]:
US_cases_c[US_cases_c[['date','county', 'Abbreviation']].duplicated()]

Empty DataFrame
Columns: [date, county, state, Abbreviation, cases, deaths]
Index: []

# Geographical boundaries of US counties data

In [154]:
US_county_geo = pd.read_json('Data/gz_2010_us_050_00_500k.json')
US_county_geo.head()

type                                           features
0  FeatureCollection  {'type': 'Feature', 'properties': {'GEO_ID': '...
1  FeatureCollection  {'type': 'Feature', 'properties': {'GEO_ID': '...
2  FeatureCollection  {'type': 'Feature', 'properties': {'GEO_ID': '...
3  FeatureCollection  {'type': 'Feature', 'properties': {'GEO_ID': '...
4  FeatureCollection  {'type': 'Feature', 'properties': {'GEO_ID': '...

__As we did before, let's check to see if county names or county ids don't match our original US_pop data__

In [155]:
counties = []
for i in range(len(US_county_geo['features'])):
    counties.append(US_county_geo['features'][i]['properties']['NAME'])

counties_id = []
for i in range(len(US_county_geo['features'])):
    counties_id.append(US_county_geo['features'][i]['properties']['GEO_ID'])


__We find that there are many names that don't match, some due to non alpha chars__

In [156]:
for i in counties:
    if i not in US_pop['County'].values:
        print(i)

Aleutians East
Aleutians West
Anchorage
Bethel
Bristol Bay
Denali
Dillingham
Fairbanks North Star
Hoonah-Angoon
Kenai Peninsula
Ketchikan Gateway
Lake and Peninsula
Matanuska-Susitna
Nome
North Slope
Northwest Arctic
Sitka
Southeast Fairbanks
Wade Hampton
Wrangell
Yukon-Koyukuk
Petersburg
Prince of Wales-Hyder
Skagway
Valdez-Cordova
Yakutat
Haines
Kodiak Island
Iberville
Natchitoches
Rapides
St. Bernard
St. James
St. John the Baptist
St. Mary
West Baton Rouge
West Feliciana
Plaquemines
Acadia
Ascension
Assumption
Avoyelles
Bienville
Bossier
Catahoula
Concordia
De Soto
East Baton Rouge
East Carroll
East Feliciana
Evangeline
Iberia
Lafourche
Morehouse
Pointe Coupee
St. Helena
St. Landry
St. Martin
St. Tammany
Tangipahoa
Tensas
Terrebonne
West Carroll
Winn
Beauregard
Calcasieu
Do�a Ana
Colonial Heights
Falls Church
Staunton
Winchester
Alexandria
Charlottesville
Chesapeake
Danville
Emporia
Fredericksburg
Galax
Harrisonburg
Hopewell
Lynchburg
Manassas
Manassas Park
Martinsville
Newport News

__However, here we find that only 3 don't match when we compare using county ids. So we will stick to the county ids when we want to plot county boundaries on maps__

In [157]:
for i in counties_id:
    if i not in US_pop['Id'].values:
        print(i)

0500000US02270
0500000US46113
0500000US51515


In [158]:
for i in range(len(US_county_geo['features'])):  
    if US_county_geo['features'][i]['properties']['GEO_ID'] == '0500000US02270':
        print(US_county_geo['features'][i]['properties']['NAME'])

Wade Hampton


In [159]:
US_pop[US_pop['County'].str.contains('Wade Hampton')]

Empty DataFrame
Columns: [Id, Id2, County, State, Population Estimate 2018]
Index: []

# Geographical locations of US counties data

In [160]:
US_county_geo_loc = pd.read_csv('Data/2019_Gaz_counties_national.txt', sep='\t')
US_county_geo_loc

USPS  GEOID  ANSICODE                 NAME       ALAND      AWATER  \
0      AL   1001    161526       Autauga County  1539602137    25706961   
1      AL   1003    161527       Baldwin County  4117621645  1132980868   
2      AL   1005    161528       Barbour County  2292160139    50523213   
3      AL   1007    161529          Bibb County  1612167476     9602089   
4      AL   1009    161530        Blount County  1670103866    15015467   
...   ...    ...       ...                  ...         ...         ...   
3215   PR  72145   1804553  Vega Baja Municipio   118745077    57827590   
3216   PR  72147   1804554    Vieques Municipio   131541395   552192819   
3217   PR  72149   1804555   Villalba Municipio    92298569     3622639   
3218   PR  72151   1804556    Yabucoa Municipio   143005186    72592516   
3219   PR  72153   1804557      Yauco Municipio   175371927     1625259   

      ALAND_SQMI  AWATER_SQMI   INTPTLAT  \
0        594.444        9.926  32.532237   
1       1589.823      437.446  30.659218   
2        885.008       19.507  31.870253   
3        622.461        3.707  33.015893   
4        644.831        5.798  33.977358   
...          ...          ...        ...   
3215      45.848       22.327  18.455128   
3216      50.788      213.203  18.125418   
3217      35.637        1.399  18.130718   
3218      55.215       28.028  18.059858   
3219      67.711        0.628  18.085669   

      INTPTLONG                                                                                                                 
0                                            -86.646440                                                                         
1                                            -87.746067                                                                         
2                                            -85.405104                                                                         
3                                            -87.127148                                                                         
4                                            -86.566440                                                                         
...                                                 ...                                                                         
3215                                         -66.397883                                                                         
3216                                         -65.432474                                                                         
3217                                         -66.472244                                                                         
3218                                         -65.859871                                                                         
3219                                         -66.857901                                                                         

[3220 rows x 10 columns]

__Drop unnecessary columns and rename others__

In [161]:
US_county_geo_loc.drop(['USPS', 'ANSICODE', 'NAME', 'ALAND', 'AWATER', 'ALAND_SQMI', 'AWATER_SQMI'], axis = 1, inplace=True)

In [162]:
US_county_geo_loc.rename(columns = {US_county_geo_loc.columns[1] : 'Latitude', US_county_geo_loc.columns[2] : 'Longitude'}, inplace = True)

__Check using the county id that all of them (3220) are in the original US_pop (3220) data, and they are indeed__

In [163]:
for i in US_pop['Id2'].values:
    if i not in US_county_geo_loc['GEOID'].values:
        print(i)    

__Now we join the population data with the geographical locations of the counties__

In [164]:
US_pop_c = US_pop_c.join(US_county_geo_loc.set_index(['GEOID']), on = ['Id2'])
US_pop_c

Id    Id2      County    State  Population Estimate 2018  \
0     0500000US01001   1001     Autauga  Alabama                     55601   
1     0500000US01003   1003     Baldwin  Alabama                    218022   
2     0500000US01005   1005     Barbour  Alabama                     24881   
3     0500000US01007   1007        Bibb  Alabama                     22400   
4     0500000US01009   1009      Blount  Alabama                     57840   
...              ...    ...         ...      ...                       ...   
3137  0500000US56037  56037  Sweetwater  Wyoming                     43051   
3138  0500000US56039  56039       Teton  Wyoming                     23081   
3139  0500000US56041  56041       Uinta  Wyoming                     20299   
3140  0500000US56043  56043    Washakie  Wyoming                      7885   
3141  0500000US56045  56045      Weston  Wyoming                      6967   

     Abbreviation   Latitude   Longitude  
0              AL  32.532237  -86.646440  
1              AL  30.659218  -87.746067  
2              AL  31.870253  -85.405104  
3              AL  33.015893  -87.127148  
4              AL  33.977358  -86.566440  
...           ...        ...         ...  
3137           WY  41.660328 -108.875677  
3138           WY  44.048662 -110.426087  
3139           WY  41.284726 -110.558947  
3140           WY  43.878831 -107.669052  
3141           WY  43.846213 -104.570020  

[3142 rows x 8 columns]

__Finally we join the previous dataframe to the hospital capacity dataframe. Again, we don't have hospital data for all the 3142 counties in the previous dataframe__

In [165]:
US_hosp_pop_c = US_hosp_pop_c.join(US_pop_c.set_index(['Id', 'Id2', 'Population Estimate 2018']), on = ['Id', 'Id2', 'Population Estimate 2018'])
US_hosp_pop_c

Id      Id2  Population Estimate 2018  NUM_LICENSED_BEDS  \
0     0500000US01001   1001.0                   55601.0               85.0   
1     0500000US01003   1003.0                  218022.0              386.0   
2     0500000US01005   1005.0                   24881.0               74.0   
3     0500000US01007   1007.0                   22400.0               35.0   
4     0500000US01009   1009.0                   57840.0               25.0   
...              ...      ...                       ...                ...   
2476  0500000US56037  56037.0                   43051.0              115.0   
2477  0500000US56039  56039.0                   23081.0               48.0   
2478  0500000US56041  56041.0                   20299.0              225.0   
2479  0500000US56043  56043.0                    7885.0               18.0   
2480  0500000US56045  56045.0                    6967.0               12.0   

      NUM_STAFFED_BEDS  NUM_ICU_BEDS  ADULT_ICU_BEDS  PEDI_ICU_BEDS  \
0                 55.0             6               6            0.0   
1                362.0            51              51            0.0   
2                 30.0             5               5            0.0   
3                 25.0             4               4            0.0   
4                 25.0             6               6            0.0   
...                ...           ...             ...            ...   
2476              74.0            13              13            0.0   
2477              48.0             6               6            0.0   
2478             105.0             6               6            0.0   
2479              18.0             3               3            0.0   
2480              12.0             2               2            0.0   

      BED_UTILIZATION  Potential_Increase_In_Bed_Capac  AVG_VENTILATOR_USAGE  \
0            0.623113                               30                   2.0   
1            0.563699                               24                   8.0   
2            0.430640                               44                   2.0   
3            0.376219                               10                   1.0   
4            0.499068                                0                   2.0   
...               ...                              ...                   ...   
2476         0.108408                               41                   2.0   
2477         0.319578                                0                   2.0   
2478         0.450803                              120                   2.0   
2479         0.287976                                0                   2.0   
2480         0.599641                                0                   0.0   

      NUM_HOSPITALS      County    State Abbreviation   Latitude   Longitude  
0                 1     Autauga  Alabama           AL  32.532237  -86.646440  
1                 4     Baldwin  Alabama           AL  30.659218  -87.746067  
2                 1     Barbour  Alabama           AL  31.870253  -85.405104  
3                 1        Bibb  Alabama           AL  33.015893  -87.127148  
4                 1      Blount  Alabama           AL  33.977358  -86.566440  
...             ...         ...      ...          ...        ...         ...  
2476              2  Sweetwater  Wyoming           WY  41.660328 -108.875677  
2477              1       Teton  Wyoming           WY  44.048662 -110.426087  
2478              2       Uinta  Wyoming           WY  41.284726 -110.558947  
2479              1    Washakie  Wyoming           WY  43.878831 -107.669052  
2480              1      Weston  Wyoming           WY  43.846213 -104.570020  

[2481 rows x 17 columns]

__Rename and reorder columns__

In [166]:
US_hosp_pop_c.rename(columns={'Abbreviation' : 'state_abb', 
                              'County' : 'county',
                              'State' : 'state',
                              'Id' : 'county_id', 
                              'Id2' : 'county_id2', 
                              'Population Estimate 2018' : 'population', 
                              'NUM_LICENSED_BEDS' : 'num_licensed_beds',
                              'NUM_STAFFED_BEDS' : 'num_staffed_beds',
                              'NUM_ICU_BEDS' : 'num_icu_beds',
                              'ADULT_ICU_BEDS' : 'adult_icu_beds',
                              'PEDI_ICU_BEDS' : 'pedi_icu_beds',
                              'BED_UTILIZATION' : 'bed_utilization',
                              'Potential_Increase_In_Bed_Capac' : 'extra_beds',
                              'AVG_VENTILATOR_USAGE' : 'avg_vent_usage',
                              'NUM_HOSPITALS' : 'num_hospitals'                             
                             }, inplace=True)

__Finally this is the df_hospitals dataframe that we will save__

In [167]:
df_hospitals = US_hosp_pop_c [['county_id', 'county_id2', 'county', 'state', 'state_abb', 'num_licensed_beds',
       'num_staffed_beds', 'num_icu_beds', 'adult_icu_beds', 'pedi_icu_beds',
       'bed_utilization', 'extra_beds', 'avg_vent_usage', 'num_hospitals',
        'population', 'Latitude', 'Longitude']]
df_hospitals

county_id  county_id2      county    state state_abb  \
0     0500000US01001      1001.0     Autauga  Alabama        AL   
1     0500000US01003      1003.0     Baldwin  Alabama        AL   
2     0500000US01005      1005.0     Barbour  Alabama        AL   
3     0500000US01007      1007.0        Bibb  Alabama        AL   
4     0500000US01009      1009.0      Blount  Alabama        AL   
...              ...         ...         ...      ...       ...   
2476  0500000US56037     56037.0  Sweetwater  Wyoming        WY   
2477  0500000US56039     56039.0       Teton  Wyoming        WY   
2478  0500000US56041     56041.0       Uinta  Wyoming        WY   
2479  0500000US56043     56043.0    Washakie  Wyoming        WY   
2480  0500000US56045     56045.0      Weston  Wyoming        WY   

      num_licensed_beds  num_staffed_beds  num_icu_beds  adult_icu_beds  \
0                  85.0              55.0             6               6   
1                 386.0             362.0            51              51   
2                  74.0              30.0             5               5   
3                  35.0              25.0             4               4   
4                  25.0              25.0             6               6   
...                 ...               ...           ...             ...   
2476              115.0              74.0            13              13   
2477               48.0              48.0             6               6   
2478              225.0             105.0             6               6   
2479               18.0              18.0             3               3   
2480               12.0              12.0             2               2   

      pedi_icu_beds  bed_utilization  extra_beds  avg_vent_usage  \
0               0.0         0.623113          30             2.0   
1               0.0         0.563699          24             8.0   
2               0.0         0.430640          44             2.0   
3               0.0         0.376219          10             1.0   
4               0.0         0.499068           0             2.0   
...             ...              ...         ...             ...   
2476            0.0         0.108408          41             2.0   
2477            0.0         0.319578           0             2.0   
2478            0.0         0.450803         120             2.0   
2479            0.0         0.287976           0             2.0   
2480            0.0         0.599641           0             0.0   

      num_hospitals  population   Latitude   Longitude  
0                 1     55601.0  32.532237  -86.646440  
1                 4    218022.0  30.659218  -87.746067  
2                 1     24881.0  31.870253  -85.405104  
3                 1     22400.0  33.015893  -87.127148  
4                 1     57840.0  33.977358  -86.566440  
...             ...         ...        ...         ...  
2476              2     43051.0  41.660328 -108.875677  
2477              1     23081.0  44.048662 -110.426087  
2478              2     20299.0  41.284726 -110.558947  
2479              1      7885.0  43.878831 -107.669052  
2480              1      6967.0  43.846213 -104.570020  

[2481 rows x 17 columns]

__We join the cases dataframe to the df_hospitals dataframe and then create df_cases dataframe which will save__

In [168]:
df_cases = US_cases_c.join(df_hospitals.drop('state', axis=1).set_index(['county', 'state_abb']), on = ['county', 'Abbreviation']).rename(columns = {'Abbreviation':'state_abb'})
df_cases = df_cases[['date', 'county_id', 'county_id2', 'county', 'state' , 'state_abb' , 'cases', 'deaths',
        'num_licensed_beds', 'num_staffed_beds',
       'num_icu_beds', 'adult_icu_beds', 'pedi_icu_beds', 'bed_utilization',
       'extra_beds', 'avg_vent_usage', 'num_hospitals', 'population',
       'Latitude', 'Longitude']]
df_cases.dropna(inplace=True)
df_cases.reset_index(drop=True, inplace=True)
df_cases

date       county_id  county_id2     county       state  \
0      2020-01-21  0500000US53061     53061.0  Snohomish  Washington   
1      2020-01-22  0500000US53061     53061.0  Snohomish  Washington   
2      2020-01-23  0500000US53061     53061.0  Snohomish  Washington   
3      2020-01-24  0500000US17031     17031.0       Cook    Illinois   
4      2020-01-24  0500000US53061     53061.0  Snohomish  Washington   
...           ...             ...         ...        ...         ...   
68390  2020-04-22  0500000US51199     51199.0       York    Virginia   
68391  2020-04-22  0500000US48503     48503.0      Young       Texas   
68392  2020-04-22  0500000US06115      6115.0       Yuba  California   
68393  2020-04-22  0500000US04027      4027.0       Yuma     Arizona   
68394  2020-04-22  0500000US08125      8125.0       Yuma    Colorado   

      state_abb  cases  deaths  num_licensed_beds  num_staffed_beds  \
0            WA      1       0              993.0             863.0   
1            WA      1       0              993.0             863.0   
2            WA      1       0              993.0             863.0   
3            IL      1       0            18546.0           15382.0   
4            WA      1       0              993.0             863.0   
...         ...    ...     ...                ...               ...   
68390        VA     43       1              195.0             189.0   
68391        TX      4       1               50.0              50.0   
68392        CA     16       1              221.0             149.0   
68393        AZ     28       1              456.0             447.0   
68394        CO      4       0               30.0              30.0   

       num_icu_beds  adult_icu_beds  pedi_icu_beds  bed_utilization  \
0              72.0            72.0           38.0         0.566322   
1              72.0            72.0           38.0         0.566322   
2              72.0            72.0           38.0         0.566322   
3            2021.0          2021.0          651.0         0.501626   
4              72.0            72.0           38.0         0.566322   
...             ...             ...            ...              ...   
68390          16.0            16.0            0.0         0.609943   
68391           8.0             8.0            0.0         0.279945   
68392          24.0            24.0            6.0         0.707348   
68393          42.0            42.0           20.0         0.585519   
68394           6.0             6.0            0.0         0.185702   

       extra_beds  avg_vent_usage  num_hospitals  population   Latitude  \
0           130.0            25.0            5.0    814901.0  48.070464   
1           130.0            25.0            5.0    814901.0  48.070464   
2           130.0            25.0            5.0    814901.0  48.070464   
3          3164.0           455.0           68.0   5180493.0  41.894294   
4           130.0            25.0            5.0    814901.0  48.070464   
...           ...             ...            ...         ...        ...   
68390         6.0             5.0            2.0     67846.0  37.219081   
68391         0.0             3.0            2.0     18045.0  33.158778   
68392        72.0            12.0            1.0     78041.0  39.270130   
68393         9.0             7.0            2.0    212128.0  32.773942   
68394         0.0             1.0            2.0     10020.0  40.000770   

        Longitude  
0     -121.936784  
1     -121.936784  
2     -121.936784  
3      -87.645455  
4     -121.936784  
...           ...  
68390  -76.561648  
68391  -98.678390  
68392 -121.344259  
68393 -113.910905  
68394 -102.421649  

[68395 rows x 20 columns]

__The following dataframe will contain all the cases even those in counties without known hospital capacity__

In [169]:
df_cases_all = US_cases_c.join(US_pop_c.set_index(['County', 'Abbreviation']), on = ['county', 'Abbreviation']).drop('State', axis = 1)
df_cases_all.rename(columns = {'Abbreviation' : 'state_abb', 'Id' : 'county_id', 'Id2' : 'county_id2', 'Population Estimate 2018' : 'population'}, inplace = True)
df_cases_all = df_cases_all[['date', 'county_id', 'county_id2', 'county', 'state' , 'state_abb' , 'cases', 'deaths', 'population', 'Latitude', 'Longitude']]
df_cases_all

date       county_id  county_id2                     county  \
0      2020-01-21  0500000US53061       53061                  Snohomish   
1      2020-01-22  0500000US53061       53061                  Snohomish   
2      2020-01-23  0500000US53061       53061                  Snohomish   
3      2020-01-24  0500000US17031       17031                       Cook   
4      2020-01-24  0500000US53061       53061                  Snohomish   
...           ...             ...         ...                        ...   
80402  2020-04-22  0500000US02290        2290  Yukon-Koyukuk Census Area   
80403  2020-04-22  0500000US04027        4027                       Yuma   
80404  2020-04-22  0500000US08125        8125                       Yuma   
80405  2020-04-22  0500000US48505       48505                     Zapata   
80406  2020-04-22  0500000US48507       48507                     Zavala   

            state state_abb  cases  deaths  population   Latitude   Longitude  
0      Washington        WA      1       0      814901  48.070464 -121.936784  
1      Washington        WA      1       0      814901  48.070464 -121.936784  
2      Washington        WA      1       0      814901  48.070464 -121.936784  
3        Illinois        IL      1       0     5180493  41.894294  -87.645455  
4      Washington        WA      1       0      814901  48.070464 -121.936784  
...           ...       ...    ...     ...         ...        ...         ...  
80402      Alaska        AK      1       0        5327  65.375727 -151.577850  
80403     Arizona        AZ     28       1      212128  32.773942 -113.910905  
80404    Colorado        CO      4       0       10020  40.000770 -102.421649  
80405       Texas        TX      7       0       14190  26.996981  -99.182603  
80406       Texas        TX      1       0       11983  28.864652  -99.759830  

[80407 rows x 11 columns]

__Saving the three above dataframes__

In [170]:
df_hospitals.count()

county_id            2481
county_id2           2481
county               2481
state                2481
state_abb            2481
num_licensed_beds    2481
num_staffed_beds     2481
num_icu_beds         2481
adult_icu_beds       2481
pedi_icu_beds        2481
bed_utilization      2481
extra_beds           2481
avg_vent_usage       2481
num_hospitals        2481
population           2481
Latitude             2481
Longitude            2481
dtype: int64

In [171]:
df_cases.count()

date                 68395
county_id            68395
county_id2           68395
county               68395
state                68395
state_abb            68395
cases                68395
deaths               68395
num_licensed_beds    68395
num_staffed_beds     68395
num_icu_beds         68395
adult_icu_beds       68395
pedi_icu_beds        68395
bed_utilization      68395
extra_beds           68395
avg_vent_usage       68395
num_hospitals        68395
population           68395
Latitude             68395
Longitude            68395
dtype: int64

In [172]:
df_cases_all.count()

date          80407
county_id     80407
county_id2    80407
county        80407
state         80407
state_abb     80407
cases         80407
deaths        80407
population    80407
Latitude      80407
Longitude     80407
dtype: int64

In [173]:
df_hospitals.to_csv('Data/df_hospitals.csv', index=False)

In [174]:
df_cases.to_csv('Data/df_cases.csv', index=False)

In [175]:
df_cases_all.to_csv('Data/df_cases_all.csv', index=False)

In [2]:
#%history -g -f filename